vault/identity_store_oidc.go

```diff
@@ -556,46 +575,69 @@ func (i *IdentityStore) pathOIDCReadKey(ctx context.Context, req *logical.Reques
 	}, nil
 }
 
-// handleOIDCDeleteKey is used to delete a key
-func (i *IdentityStore) pathOIDCDeleteKey(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
-	ns, err := namespace.FromContext(ctx)
-	if err != nil {
-		return nil, err
-	}
-
-	targetKeyName := d.Get("name").(string)
-
-	i.oidcLock.Lock()
-
-	// it is an error to delete a key that is actively referenced by a role
+// getRolesReferencingTargetKeyName returns a map of role names to roles referenced by targetKeyName.
+// Note: this is not threadsafe. It is to be called with Lock already held.
+func (i *IdentityStore) getRolesReferencingTargetKeyName(ctx context.Context, req *logical.Request, targetKeyName string) (map[string]role, error) {
 	roleNames, err := req.Storage.List(ctx, roleConfigPath)
 	if err != nil {
-		i.oidcLock.Unlock()
 		return nil, err
 	}
 
-	var role *role
-	rolesReferencingTargetKeyName := make([]string, 0)
+	var tempRole role
+	rolesReferencingTargetKeyName := make(map[string]role)
 	for _, roleName := range roleNames {
 		entry, err := req.Storage.Get(ctx, roleConfigPath+roleName)
 		if err != nil {
-			i.oidcLock.Unlock()
 			return nil, err
 		}
 		if entry != nil {
-			if err := entry.DecodeJSON(&role); err != nil {
-				i.oidcLock.Unlock()
+			if err := entry.DecodeJSON(&tempRole); err != nil {
 				return nil, err
 			}
-			if role.Key == targetKeyName {
-				rolesReferencingTargetKeyName = append(rolesReferencingTargetKeyName, roleName)
+			if tempRole.Key == targetKeyName {
+				rolesReferencingTargetKeyName[roleName] = tempRole
 			}
 		}
 	}
 
-	if len(rolesReferencingTargetKeyName) > 0 {
+	return rolesReferencingTargetKeyName, nil
+}
+
+// getRoleNamesReferencingTargetKeyName returns a slice of strings of role
+// names referenced by targetKeyName.
+// Note: this is not threadsafe. It is to be called with Lock already held.
+func (i *IdentityStore) getRoleNamesReferencingTargetKeyName(ctx context.Context, req *logical.Request, targetKeyName string) ([]string, error) {
+	rolesReferencingTargetKeyName, err := i.getRolesReferencingTargetKeyName(ctx, req, targetKeyName)
+	if err != nil {
+		return nil, err
+	}
+
+	var names []string
+	for key, _ := range rolesReferencingTargetKeyName {
+		names = append(names, key)
+	}
+	return names, nil
```

Comment:
```suggestion
	}
	sort.Strings(names)
	return names, nil
```

Wondering if we should return it sorted to make the result consistent.